### ISSUE TITLE
`__before_ready()` is not called if there's no `#[godot_api]` trait impl

### ISSUE THREAD

**Issue by Houtamelo** [godot-rust/gdext#913]
Some features we implemented require __before_ready() to work:
- OnReady<T> initialization
- Rpcs configuration

Consider the class:
```rust
#[derive(GodotClass)]
#[class(init, base = Node)]
struct MyClass;
```

If it does not have a IClass impl:
```rust
#[godot_api]
impl INode for MyClass {}
```

Then `__before_ready()` will never be run, which makes the features above fail "silently".

---

`__before_ready()` is only called in code generated by the macro expansion of the #[godot_api] trait impl:
```rust
impl ::godot::obj::cap::ImplementsGodotVirtual for OwnerDataApi {
	fn __virtual_call(name: &str) -> ::godot::sys::GDExtensionClassCallVirtual {
		use ::godot::obj::UserClass   as _;
		match name {
			"_ready" => {
				use ::godot::sys;
				type Sig = ((),);
				unsafe extern "C" fn virtual_fn(instance_ptr: sys::GDExtensionClassInstancePtr, args_ptr: *const sys::GDExtensionConstTypePtr, ret: sys::GDExtensionTypePtr) {
					let call_ctx = ::godot::meta::CallContext::func("OwnerDataApi", "ready");
					let _success = ::godot::private::handle_ptrcall_panic(&call_ctx, || <Sig as ::godot::meta::PtrcallSignatureTuple>::in_ptrcall(instance_ptr, &call_ctx, args_ptr, ret, |instance_ptr, params| {
						let () = params;
						let storage = unsafe { ::godot::private::as_storage::<OwnerDataApi>(instance_ptr) };
						let mut instance = ::godot::private::Storage::get_mut(storage);
						instance.__before_ready(); // <---------------------------
					}, sys::PtrcallType::Virtual, ));
				}
				Some(virtual_fn)
			}
			_ => None,
		}
	}
}
```

A temporary fix is to simply add an empty impl if you don't need to override it:
```rust
#[godot_api]
impl INode for MyClass {}
```

**Comment by lilizoey** [comment]
we could probably just add a variant of `You_forgot_the_attribute__godot_api` for the itrait impl too.

**Comment by Bromeon** [comment]
@Houtamelo do you have a minimal test case that demonstrates the problem?

Because this one seems to work:
https://github.com/godot-rust/gdext/blob/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff/itest/rust/src/object_tests/onready_test.rs#L95-L110

**Comment by Houtamelo** [comment]
> [@Houtamelo](https://github.com/Houtamelo) do you have a minimal test case that demonstrates the problem?
> 
> Because this one seems to work:
> 
> [gdext/itest/rust/src/object_tests/onready_test.rs](https://github.com/godot-rust/gdext/blob/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff/itest/rust/src/object_tests/onready_test.rs#L95-L110)
> 
> Lines 95 to 110 in [f67a1ae](/godot-rust/gdext/commit/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff)
>  #[itest] 
>  fn onready_lifecycle_without_impl() { 
>      let mut obj = OnReadyWithoutImpl::create(); 
>   
>      obj.notify(NodeNotification::READY); 
>   
>      { 
>          let mut obj = obj.bind_mut(); 
>          assert_eq!(*obj.auto, 44); 
>   
>          *obj.auto = 55; 
>          assert_eq!(*obj.auto, 55); 
>      } 
>   
>      obj.free(); 
>  }

The snippets I provided in my initial post are the minimal test case.
That being said, I reported this months ago, maybe this behavior was fixed?

**Comment by Bromeon** [comment]
> The snippets I provided in my initial post are the minimal test case.

You mentioned:

> Some features we implemented require __before_ready() to work:
>  * OnReady initialization
>  * Rpcs configuration
> 
> Consider the class:
> 
> ```rs
> #[derive(GodotClass)]
> #[class(init, base = Node)]
> struct MyClass;
> ```
> 
> If it does not have a IClass impl [...] Then `__before_ready()` will never be run, which makes the features above fail "silently".

However, that class has neither `#[rpc]` nor `OnReady` fields, so it doesn't demonstrate the issue. Adding `OnReady` alone isn't enough because it also needs a base field.

---

In my last post, I linked to a test case which uses this setup and works:
(note that the `impl` block has no `#[godot_api]` macro)

https://github.com/godot-rust/gdext/blob/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff/itest/rust/src/object_tests/onready_test.rs#L214-L231

 So I assume this issue has been solved in the meantime. If there's a way to still reproduce it, let me know.

**Comment by Houtamelo** [comment]
> > The snippets I provided in my initial post are the minimal test case.
> 
> You mentioned:
> 
> > Some features we implemented require __before_ready() to work:
> > 
> > * OnReady initialization
> > * Rpcs configuration
> > 
> > Consider the class:
> > #[derive(GodotClass)]
> > #[class(init, base = Node)]
> > struct MyClass;
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If it does not have a IClass impl [...] Then `__before_ready()` will never be run, which makes the features above fail "silently".
> 
> However, that class has neither `#[rpc]` nor `OnReady` fields, so it doesn't demonstrate the issue. Adding `OnReady` alone isn't enough because it also needs a base field.
> 
> In my last post, I linked to a test case which uses this setup and works: (note that the `impl` block has no `#[godot_api]` macro)
> 
> [gdext/itest/rust/src/object_tests/onready_test.rs](https://github.com/godot-rust/gdext/blob/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff/itest/rust/src/object_tests/onready_test.rs#L214-L231)
> 
> Lines 214 to 231 in [f67a1ae](/godot-rust/gdext/commit/f67a1ae43b1e06cd9a58ccb9d6330bf06c6e25ff)
>  #[derive(GodotClass)] 
>  #[class(no_init, base=Node)] 
>  struct OnReadyWithoutImpl { 
>      base: Base<Node>, 
>      auto: OnReady<i32>, 
>      // No manual one, since those cannot be initialized without a ready() override. 
>      // (Technically they _can_ at the moment, but in the future we might ensure initialization after ready, so this is not a supported workflow). 
>  } 
>   
>  // Rust-only impl, no proc macros. 
>  impl OnReadyWithoutImpl { 
>      fn create() -> Gd<OnReadyWithoutImpl> { 
>          Gd::from_init_fn(|base| Self { 
>              base, 
>              auto: OnReady::new(|| 44), 
>          }) 
>      } 
>  } 
> 
> So I assume this issue has been solved in the meantime. If there's a way to still reproduce it, let me know.

You're absolutely correct, I glanced over my (old) initial post too quickly and didn't realize it was simplified, the project that used the rpc annotations has been put in the shelf for a while so I can't easily test this again, I'll ask to re-open the issue if I run into this again in the future
